0001-kaa-protocol/README.md

```diff
@@ -142,16 +120,97 @@ Thus, the resource paths for them will start with `kp1/<appversion_name>/<extens
+### Request ID
+CoAP message has a built-in Token which is used to match responses to requests.
+
+MQTT PUBLISH packet includes a Packet Identifier.
+However, Packet Identifiers are not stable; that means, when a PUBLISH packet is re-transmitted by MQTT broker or MQTT gateway, the Packet Identifier may be changed.
+Thus, MQTT Packet Identifiers are not suitable for matching responses to requests.
+
+KP over MQTT MUST NOT use Packet Identifiers to match responses to requests.
+
+KP can not make assumptions about payload format, thus it cannot define the format for passing Request IDs within the message payload.
+Thus, Request IDs are embedded within the topic name.
+
+In case the last topic level in MQTT topic is a positive integer number, it MUST be interpreted as a Request ID.
+
+For example, the request sent to `<endpoint_token>/update/42` topic is a request to `<endpoint_token>/update` Resource Path with Request ID `42`.
+
+If Request ID is not present in the MQTT topic, the response SHOULD NOT be published.
+
+> Note: Request ID in the topic name is MQTT-specific and is not counted toward Resource Path.
+
+### Response topic
+For CoAP, request and response semantics are carried in CoAP messages.
+
+MQTT does not have a request/response notion, so a separate MQTT topic is introduced to differentiate responses.
+
+Successful responses over MQTT MUST be published to the topic constructed by appending `/status` suffix to the request topic.
+
+Error responses over MQTT MUST be published to the topic constructed by appending `/error` suffix to the request topic.
+
+For example, request sent to `<endpoint_token>/json/42` has the following success response topic: `<endpoint_token>/json/42/status`; the error response topic is `<endpoint_token>/json/42/error`.
+
+> Note: response type suffix is MQTT-specific and is not counted toward Resource Path.
+
+### Error response format
+Error responses published over MQTT MUST be UTF-8 encoded JSON object with the scheme defined in [error-response.schema.json](./error-response.schema.json) file:
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+  "description": "Kaa Protocol error response",
+  "type": "object",
+  "properties": {
+    "statusCode": {
+      "type": "integer",
+      "description": "Status code based on HTTP status codes"
+    },
+    "reasonPhrase": {
+      "type": "string",
+      "description": "A human-readable string explaining the cause of an error"
+    }
+  },
+  "required": [
+    "statusCode",
+    "reasonPhrase"
+  ],
+  "additionalProperties": false
+}
 ```
 
+For CoAP, error responses should leverage built-in CoAP capabilities:
+- Status Code should map to CoAP Response Codes ([RFC 7252, section 5.9](https://tools.ietf.org/html/rfc7252#section-5.9)).
+- Reason Phrase should be transfered as CoAP Diagnostic Payload ([RFC 7252, section 5.5.2](https://tools.ietf.org/html/rfc7252#section-5.5.2)).
+
+Extensions MAY reuse the same format for successful responses.
```

Comment:
The format above is only defined for error responses (ones that come to `/error`). This sentence encourages extension authors to reuse the same format for successful responses too.